### ISSUE TITLE
gradients inside gradient checkpoint

### ISSUE THREAD

**Issue by prolearner** [pytorch/pytorch#32005]
## 🚀 Feature
allowing the use of `torch.autorgrad.grad` and `loss.backward` inside `torch.utils.checkpoint.checkpoint`

## Motivation
Enclosing the entire training steps inside a checkpoint could improve significantly the memory cost for the  gradient computation in large scale hyperparameter optimization (HO) and meta-learning algorithms (ML) based on unrolled differentiation. Here you have to backpropagate over the entire training procedure and doing that naively with pytorch has a really high cost in memory because it records all the activations and all the weights for all the training steps. Using checkpointing you could just store the weights for all the training steps and recompute the activation and the weight update operation (which include the gradient of the loss) during the backward pass.

## Pitch
I tried to enclosed backward operations inside a checkpoint and got an error on the gradient computation. I think it will be nice if you could do that so that you could see the benefit of checkpointing for HO and ML which can be really demanding in terms of memory.

## Alternatives
The alternative to use checkpointing is to implement a custom backward pass for these use cases exploiting the particular structure of the problem to save memory. I personally think that the requested feature provides a much simpler solution.

**Comment by albanD** [comment]
Hi,

I think there are few different things in your question:

## Support autograd.grad with checkpoint.
This is not actually possible. When running `autograd.grad`, we need to know where the inputs are. But we don't know which inputs where used in the checkpoint (in particular for the inputs that are implicitly captured by your function). So we cannot do this discovery.
`autograd.backward()` is allowed because **all** edges in the graph need to be explored because we created the graph this way.

## Support loss.backward inside the checkpoint
This is already allowed.
You need to be careful though that your function does not have side effect (which is pretty hard when using autograd.backaward()`):
```python

import torch
from torch import autograd
from torch.utils import checkpoint

from torchviz import make_dot

net = torch.nn.Linear(3, 3)

def f(inp):
    # Do not change the original input
    inp = inp.detach()
    # Preserve net gradients
    orig_grad = []
    for p in net.parameters():
        orig_grad.insert(0, p.grad)
        p.grad = None

    # Your code
    with torch.enable_grad():
        embed = net(inp)

        embed.retain_grad()
        loss = (embed ** 2).sum()

        loss.backward(retain_graph=True)

    # Restore net gradients
    for p in net.parameters():
        p.grad = orig_grad.pop()

    # Rescale embed by its gradient norm as an example
    return embed * embed.grad

inp = torch.ones(1, 3, requires_grad=True)

loss = checkpoint.checkpoint(f, inp).sum()
net.zero_grad()
loss.backward()
for p in net.parameters():
    print("param with checkpoint")
    print(p, p.grad)

loss = f(inp).sum()
net.zero_grad()
loss.backward()
for p in net.parameters():
    print("param without checkpoint")
    print("param")
    print(p, p.grad)
```

## The problem of doing memory-efficient meta learning

Here you will encounter a couple of problem.
- The main one being that `torch.optim` optimizers works outside of the autograd by using `.detach()` or `torch.no_grad()`. Meaning that the updates they do will not be registered. So you cannot use these in a differentiable manner. This may be changed in the future cc @vincentqb ?
- The second one is that as mentioned in the point above, you can only use locally created graphs. But in your case, the local graph should be linked to the global one. So you will need to hack around this limitation by using automatically captures Tensors.
- Finally, the limitation to use autograd.backward means that at every iteration, you will backward through all the graph that was created up to now. This is going to make the training really really slow.

The good thing for this particular application is that your function that you want to checkpoint **never** automatically captures stuff outside of its inputs (because you want to specify the current state of the weights, not use the ones in your model).
Given this, you can have a different checkpoint function that is more versatile. The following code learns what is the best lr for my dummy task such that the loss after 10 epochs is as low as possible.
```python
import torch
from torch import autograd
from torchviz import make_dot

torch.manual_seed(123)
init_param = torch.rand(2)
inp = torch.rand(2)
label = torch.rand(2)

class NoCaptureCheckpoint(autograd.Function):
    @staticmethod
    def forward(ctx, fn, *args):
        ctx.save_for_backward(*args)
        ctx.fn = fn
        with torch.no_grad():
            outputs = fn(*args)
        return outputs

    @staticmethod
    def backward(ctx, *grads):
        args = ctx.saved_tensors
        with torch.enable_grad():
            outputs = ctx.fn(*args)
        return (None,) + autograd.grad(outputs, args, grads)

my_checkpoint = NoCaptureCheckpoint.apply

def compute_loss(inp, label, param):
    x = inp * param
    return (x - label).pow(2).abs().sum()

def one_update(param, lr):
    # This function is NOT allowed to use anything else but param and lr as differentiable inputs

    # Only needed because this might run in `no_grad` block (due to checkpoint)
    with torch.enable_grad():
        loss = compute_loss(inp, label, param)

    grad = autograd.grad(loss, param, create_graph=True)[0]

    new_param = param - lr * grad

    return loss, new_param

def eval_lr_loss(param_lr, with_checkpoint):
    param = init_param.clone().requires_grad_()
    for e in range(10):
        tmp = param_lr.clone()

        if with_checkpoint:
            loss, param = my_checkpoint(one_update, param, tmp)
        else:
            loss, param = one_update(param, tmp)
    return loss


def run(with_checkpoint):
    print("Running with with_checkpoint:{}".format(with_checkpoint))
    param_lr = torch.tensor(.01, requires_grad=True)

    for epoch in range(10):
        lr_loss = eval_lr_loss(param_lr, with_checkpoint)
        # Minimize the final loss
        lr_grad = autograd.grad(lr_loss, param_lr)[0]
        # Could use lr_loss.backward() here but that would compute grads for all the inner parameters. This is more efficient.

        with torch.no_grad():
            param_lr -= 0.01 * lr_grad

        print("Curr loss: {}, Current solution:{}".format(lr_loss.item(), param_lr.item()))


run(False)
run(True)

```

This version of checkpoint is not going to integrate well with `torch.nn` though.
In particular, if you want to use a model based on `torch.nn` here. You will need to pass the parameters as a `list` to `compute_loss`. And then set all the weights of the model (we can discuss that further if you want).
Also you will still have to implement at least the inner optimizer by hand. The outer one can be from `torch.optim`, but that would mean using `lr_loss.backward()` which is going to be inefficient in this particular case.

Would that be good for you?

**Comment by vincentqb** [comment]
> * The main one being that `torch.optim` optimizers works outside of the autograd by using `.detach()` or `torch.no_grad()`. Meaning that the updates they do will not be registered. So you cannot use these in a differentiable manner. This may be changed in the future cc @vincentqb ?

Yes, it's on the wish list :)
